In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing product.csv

In [2]:
df = pd.read_csv(r'/Users/lupeguevara/Documents/Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [3]:
df_prods = pd.read_csv(r'/Users/lupeguevara/Documents/Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [4]:
path = r'/Users/lupeguevara/Documents/Instacart Basket Analysis'

In [5]:
path

'/Users/lupeguevara/Documents/Instacart Basket Analysis'

# Importing orders.csv

In [6]:
df_ords = pd.read_csv(r'/Users/lupeguevara/Documents/Instacart Basket Analysis/02 Data/Original Data/orders.csv', index_col = False)

In [7]:
path = r'/Users/lupeguevara/Documents/Instacart Basket Analysis'

In [8]:
path

'/Users/lupeguevara/Documents/Instacart Basket Analysis'

# Importing orders_wrangled.csv

In [9]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'),index_col = False)

In [10]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [11]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [12]:
df.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [13]:
# Create a dataframe
df_test = pd.DataFrame()

In [14]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [15]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [16]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
df_test['mix'] = df_test['mix'].astype('str')

In [18]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [19]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [20]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

## Create subset 

In [21]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
df.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [24]:
df_prods.shape

(49693, 5)

## Create subset

In [25]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [26]:
df_prods_clean.shape

(49677, 5)

In [27]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [29]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [30]:
df_prods_clean.shape

(49677, 5)

In [31]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

# E 4.5_Task 

### Question 2

In [33]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Answer: 
Upon analyzing the dataframe there apprears to be something off in the data consistency check. In particular the "days_since_prior_order" contains a strange recording of the min (0) and max (3) value appear inaccurate by reference of the percential range when 50% is valued 7 which is higher then the max. Similar to "order_numer' with a min (1) and max (1) with a inconsistent percential range when 25% is valued 5 which greater then the max. 

### Question 3

In [34]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [35]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [36]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [37]:
# Check for mix types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Answer
No mixed types in the data in df_ords.csv

## Question 5

In [38]:
# Finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Answer
After executing the code the results showed missing values in the "days_since_prior_order". The column is missing 206209 values. 

## Question 6

In [39]:
# Create new dataframe df_nan_ords
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [40]:
df_nan_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
11        2168274        2             1          2                 11   
26        1374495        3             1          1                 14   
39        3343014        4             1          6                 11   
45        2717275        5             1          3                 12   
...           ...      ...           ...        ...                ...   
3420930    969311   206205             1          4                 12   
3420934   3189322   206206             1          3                 18   
3421002   2166133   206207             1          6                 19   
3421019   2227043   206208             1          1                 15   
3421069   3154581   206209             1          3                 11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

In [41]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [42]:
df_nan_ords.shape

(206209, 6)

In [43]:
df_nan_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
NaN    206209
Name: count, dtype: int64

In [47]:
df_ords.shape

(3421083, 6)

## Answer
The amount of missing values within "days_since_prio_order" is significant. Looking at the rows in df_nan_ords match to 206209 which are the same amount of users, it quickly becomes clear that there isn’t much you can do in terms of imputation. The missing values here are number of order, which are counts. No action needed to remove data.

## Question 7

In [44]:
# Check for duplicates
df_dups = df_nan_ords[df_nan_ords.duplicated()]

### Answer
No duplicates found in df_nan_ords - no action needed. 

In [45]:
# Exporting df_ords
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [46]:
# Exporting df_prods
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))